audbackend/core/artifactory.py

```diff
@@ -109,15 +112,39 @@ def _glob(
 
     def _ls(
             self,
-            path: str,
+            folder: str,
     ):
-        r"""List content of path."""
-        path = audfactory.url(
-            self.host,
-            repository=self.repository,
-            group_id=audfactory.path_to_group_id(path),
-        )
-        return [p.name for p in audfactory.path(path)]
+        r"""List all files under folder."""
+        root = self._folder(folder, '')
+
+        path = audfactory.path(root)
+        try:
+            paths = [str(x) for x in path.glob("**/*") if x.is_file()]
+        except self._non_existing_path_error:  # pragma: nocover
+            paths = []
```

Comment:
What case is this?

If the path does not exist it should raise an error.
If this covers the case of an empty repo with the `/` request, we should add a test for it.

Reply:
The empty repo case is covered in `Backend.ls()`. So yes, it's the case that the folder does not exist. We then return an empty list and an error is raised in `Backend.ls()`, this way we avoid that the implementation is responsible for raising the expected error. I now added a comment to the docstring that an empty list should be returned by `_ls()` if the folder does not exist.

Reply:
How do you differentiate an empty list returned by a non-existent folder from an empty list returned by an empty folder?

Reply:
Since there is no function to create a folder, empty folders cannot exist on the backend. In reality they can of course, but I think it still makes sense to raise always raise an error. Except for the special case of `/`, for which we return an empty list if no file exists yet in the repo.

Reply:
So in other words: a folder only exists if it contains at least one file.

Reply:
I'm not that excited about this solution as a user might find other ways to delete all files in a folder, but the folder still exists. I suppose that `backend.exist(folder)` will also return `True` then, and `backend.ls(folder)` will raise an error that it does not exist.

If we want to stay with the behavior that the backend should return a special value we might think of changing the value, e.g. to `None`.

Reply:
Another reason for always raising an error if no files are found is that this will guarantee the same behavior on every backend. E.g. assume we put the same file on backend A and B and remove it afterwards. Now assume on backend A removes the folder (or maybe a concept like folders does not even exist), but on backend B we the folder is kept. If we return an empty list if the folder exists, A will now return an empty list, but B raise an error.

Reply:
Yes, makes sense. I forgot that backends don't need support folders, and running `backend.exists()` on a folder would indeed not make any sense.

The only problematic part is that it then still might be useful to stay with our `audfactory` package as this provides an option to the user to check if folders exist on Artifactory and remove them.

Reply:
Is the behavior in `Backend.versions()` that you get `[]` if the folder is empty and a `FileNotFoundError` if the folder does not exist?

Reply:
No, currently we never raise an error but return an empty list if the file does not exist.

Reply:
I created https://github.com/audeering/audbackend/issues/61 to track the proposed change for `backend.versions()`